### PR TITLE
Add types with MonkeyType

### PR DESCRIPTION
**Pull request by scottbarnes**
Closes no issue

Feature

### Technical

This hypothetical PR is based on output from [MonkeyType](https://github.com/Instagram/MonkeyType), which can add type hints based on the types used at runtime (although in this draft it is based on the runtime of the Docker container).

I am curious if it is worth pursuing MonkeyType further. On the one hand, it looks as if it would be a pretty good way to get type information for a large part of the code base, based on the types actually used. 

On the other hand, this didn't run perfectly. There were some issues with circular imports, and something about `Couldn't find statsd_server section in config`. Those issues appear to have prevented the collection of any type information whatsoever for modules so affected. So to get more types, these errors would need to be addressed.

Further, the type hints don't follow the style we use (e.g. MonkeyType generates `Union[str, int]` instead of `str | int`), so those would need updating, though likely that could be scripted.

Additionally, adding these hints led to 65 new `mypy` errors. So even thtough MonkeyType can identify some types for us, more work would need to be done to resolve type issues that the new additions lead to.

### Stakeholders

@cclauss

### CODE REVIEW

**Comment by cclauss** [comment]
I am a fan of these changes...

> `Couldn't find statsd_server section in config`

This is a _pre-existing condition_ that I believe only happens on `localhost` and could _in theory_ be fixed by swallowing the exception or by adding `statsd_server` to the config and/or running a `statsd` server (https://graphiteapp.org) on `localhost` port `9090`.

https://github.com/internetarchive/openlibrary/blob/9b7fda26cd154239ac498e06ee513a239c233ee8/openlibrary/core/stats.py#L32
https://github.com/internetarchive/openlibrary/blob/9b7fda26cd154239ac498e06ee513a239c233ee8/conf/openlibrary.yml#L95

---

> MonkeyType generates `Union[str, int]` instead of `str | int`)

This is easily fixed with `ruff --select=UP007 --target-version=3.11 --fix my/unformatted/python_file.py`

Caution with `solr` because I believe that we still use the legacy type format on some `solr` files.  @cdrini Is that correct?

---
% `ruff rule UP007`
# non-pep604-annotation (UP007)

Derived from the **pyupgrade** linter.

Autofix is sometimes available.

Message formats:
* Use `X | Y` for type annotations

---

On the `F821 Undefined name` errors, you will need to import the classes that you use in type hints but I believe code becomes more self-documenting when this is done.

File: openlibrary/plugins/openlibrary/connection.py [hunk 1]
```diff
@@ -19,16 +22,31 @@
 class ConnectionMiddleware:
     response_type = "json"
 
-    def __init__(self, conn):
+    def __init__(
+        self, conn: Union[HybridConnection, MigrationMiddleware, MemcacheMiddleware]
```
Comment (cclauss) [inline openlibrary/plugins/openlibrary/connection.py:26]: These classes are defined further down in this same file so delay their evaluation.
```suggestion
        self, conn: Union["HybridConnection", "MigrationMiddleware", "MemcacheMiddleware"]
```
Alternative solutions would be to move this class further down in the file or...
```
        self, conn: ConnectionMiddleware
```

File: openlibrary/plugins/upstream/models.py [hunk 2]
```diff
@@ -15,21 +15,25 @@
...
 from openlibrary.utils.isbn import isbn_10_to_isbn_13, isbn_13_to_isbn_10
 from openlibrary.utils.lccn import normalize_lccn
+from infogami.infobase.client import Nothing
+from openlibrary.plugins.upstream.models import Edition
+from typing import Any, Dict, List, Optional, Union
+from web.utils import Storage
 
 
-def follow_redirect(doc):
+def follow_redirect(doc: Union[Thing, Author]) -> "Author":
```
Comment (cclauss) [inline openlibrary/plugins/upstream/models.py:36]: ???
```suggestion
def follow_redirect(doc: Union[Thing, models.Author]) -> "Author":
```
or remove the type hint if it is more work that it is worth.

Comment (scottbarnes) [reply openlibrary/plugins/upstream/models.py:36]: I have been running into the circular import thing fairly frequently, but so far with a combination of:
```
if TYPE_CHECKING:
    <import the upstream stuff>
```
and then putting quotes around the things to make them strings, such as `"Author"` has above, seems to be working.

File: openlibrary/plugins/openlibrary/connection.py (hunk 1 above)
Comment (scottbarnes) [reply openlibrary/plugins/openlibrary/connection.py:26]: I think with these the quotes are the way to go, but for some reason they only seem to work with the `Union` syntax. If I try with a pipe, there is some error that I should probably investigate.

File: openlibrary/plugins/upstream/models.py (hunk 2 above)
Comment (cclauss) [reply openlibrary/plugins/upstream/models.py:36]: I think that importing `models.Author` and `models.Edition` and defining local classes `Author` and `Edition` is better.

**Comment by scottbarnes** [comment]
I will just close this and perhaps if there is interest in continued type hinting this can be revisited.

**Pull request closed without merging**